filebeat/docs/inputs/input-kafka.asciidoc

```diff
@@ -102,6 +102,10 @@ How long to wait before retrying a failed read. Default is 2s.
 How long to wait for the minimum number of input bytes while reading. Default
 is 250ms.
 
+===== `max_processing_time`
+
+How long to wait for a message to be processed. Default is 20s.
```

Comment:
What would be the impact of this setting? What happens on timeout? Resends or data lost? Is 20s good enough, or should we rather go with some bigger value?

@faec Opinions?

Reply:
The consequences of this should be minor in our case, but I think the initial message might have been misunderstood (it's just an info message -- it means the output can't keep up with the input throughput, but doesn't necessarily mean anything is wrong, and it doesn't indicate data loss). I'm a little wary of exposing this setting since I think it will generally not do what people expect, especially with such a high default value (sarama's default is 100ms). A short summary:

When an update is requested from Kafka, sarama asks for _all_ active topics / partitions. It then passes any responses back to the respective consumers. An update iteration doesn't complete until all responses have been dispatched to the buffered consumer channels, at which point another update is requested. Because the consumer buffers are finite (`ChannelBufferSize`), this means that it's possible for slow processing on a single partition to block _all_ messages for the consumer.

This timeout setting says that if we're waiting to buffer the events for a particular partition for more than `MaxProcessingTime`, then that partition will be skipped during updates until we catch up (i.e. until all pending events from that partition have been written to their buffer channel). This allows updates for other topics / partitions to proceed while the slow one catches up. When the channel is no longer blocked, that partition is resumed.

Because our handling eventually sends all events to the same global queue, this kind of robustness is mostly irrelevant for us -- all events experience roughly the same congestion, so (in theory) we don't hit any partition-specific delays. Increasing the timeout this way will, if anything, probably cause a slight slowdown: with a low timeout, saturated queues are skipped when requesting updates, but with a 20s timeout the entire update process for _all_ queues would pause whenever _any_ of them is saturated. (This is not such a terrible outcome for us because of our global queue, but it is probably the opposite of what the original post was looking for.)

So this setting is really just a back-pressure knob with higher timeouts allowing individual saturated queues to delay global updates for longer, but we already have a lot of those and the per-partition granularity doesn't gain us much that I can see. If anything we might consider exposing `ChannelBufferSize`, which turns roughly the same knob as this  setting, but in a direction more consistent with the intuitive description of the flag (though it is still mostly redundant with our global queue settings).

Reply:
Thank you for the input. As this setting has not much of an impact, plus is a very internal setting of a library we depend on I'm rather leaning to not make it configurable. Unfortunately sarama logging support is very bare-bones. But reports like this make me wonder if we want to filter the log message and adjust the log level depending on known strings (ouch). The reported message might be better suited to be logged at 'debug' level.

Reply:
Thanks for the comments! I'm ok with closing this one if we decide so. Let me know.